File: backend/src/main/java/com/suplr/backend/controller/WhatsAppConnectionController.java
package com.suplr.backend.controller;

import com.suplr.backend.dto.WhatsAppDtos.WhatsAppConnectionRequest;
import com.suplr.backend.dto.WhatsAppDtos.WhatsAppConnectionResponse;
import com.suplr.backend.entity.Supplier;
import com.suplr.backend.entity.WhatsAppConnection;
import com.suplr.backend.repository.WhatsAppConnectionRepository;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.Optional;

@RestController
@RequestMapping("/auth/whatsapp-connection")
@RequiredArgsConstructor
public class WhatsAppConnectionController {

    private final WhatsAppConnectionRepository connectionRepository;

    @PutMapping
    public WhatsAppConnectionResponse upsert(
            @AuthenticationPrincipal Supplier supplier,
            @Valid @RequestBody WhatsAppConnectionRequest req
    ) {
        WhatsAppConnection conn = connectionRepository
                .findBySupplierId(supplier.getId())
                .orElseGet(() -> WhatsAppConnection.builder()
                        .supplierId(supplier.getId())
                        .build());

        conn.setBspEndpoint(req.bspEndpoint());
        conn.setBspApiKey(req.bspApiKey());
        conn.setPhoneNumber(req.phoneNumber());

        conn = connectionRepository.save(conn);
        return WhatsAppConnectionResponse.from(conn);
    }

    @GetMapping
    public ResponseEntity<WhatsAppConnectionResponse> get(
            @AuthenticationPrincipal Supplier supplier
    ) {
        Optional<WhatsAppConnection> conn = connectionRepository
                .findBySupplierId(supplier.getId());

        return conn.map(c -> ResponseEntity.ok(WhatsAppConnectionResponse.from(c)))
                .orElse(ResponseEntity.noContent().build());
    }
}

File: backend/src/whatsapp/whatsapp-connection.controller.ts
import {
  Controller,
  Get,
  Put,
  Body,
  UseGuards,
  HttpCode,
  HttpStatus,
} from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { WhatsAppConnection } from '../entities/whatsapp-connection.entity.js';
import { Supplier } from '../entities/supplier.entity.js';
import { JwtAuthGuard } from '../common/guards/jwt-auth.guard.js';
import { CurrentSupplier } from '../common/decorators/current-supplier.decorator.js';
import { IsString, IsUrl } from 'class-validator';

class WhatsAppConnectionRequest {
  @IsUrl() bspEndpoint!: string;
  @IsString() bspApiKey!: string;
  @IsString() phoneNumber!: string;
}

@UseGuards(JwtAuthGuard)
@Controller('auth/whatsapp-connection')
export class WhatsAppConnectionController {
  constructor(
    @InjectRepository(WhatsAppConnection)
    private readonly connRepo: Repository<WhatsAppConnection>,
  ) {}

  @Put()
  async upsert(
    @CurrentSupplier() supplier: Supplier,
    @Body() req: WhatsAppConnectionRequest,
  ) {
    let conn = await this.connRepo.findOne({
      where: { supplierId: supplier.id },
    });
    if (!conn) {
      conn = this.connRepo.create({ supplierId: supplier.id });
    }
    conn.bspEndpoint = req.bspEndpoint;
    conn.bspApiKey = req.bspApiKey;
    conn.phoneNumber = req.phoneNumber;
    conn = await this.connRepo.save(conn);
    return {
      id: conn.id,
      bspEndpoint: conn.bspEndpoint,
      phoneNumber: conn.phoneNumber,
      connectedAt: conn.connectedAt,
    };
  }

  @Get()
  async get(@CurrentSupplier() supplier: Supplier) {
    const conn = await this.connRepo.findOne({
      where: { supplierId: supplier.id },
    });
    if (!conn) return null;
    return {
      id: conn.id,
      bspEndpoint: conn.bspEndpoint,
      phoneNumber: conn.phoneNumber,
      connectedAt: conn.connectedAt,
    };
  }
}
